File: src/types/workflow.ts
/**
 * Automation & Workflow Engine — Phase 7.5 (frontend).
 *
 * These types mirror the payload shape produced by
 * `backend/app/api/v1/workflow.py` (camelCase envelopes).
 */

export type WorkflowTriggerType = "event" | "schedule" | "manual";

export type WorkflowStatus =
  | "pending"
  | "running"
  | "completed"
  | "failed"
  | "cancelled";

export type WorkflowStepStatus =
  | "pending"
  | "running"
  | "completed"
  | "failed"
  | "skipped";

export const WORKFLOW_TRIGGER_TYPES: WorkflowTriggerType[] = [
  "event",
  "schedule",
  "manual",
];

export const WORKFLOW_STATUSES: WorkflowStatus[] = [
  "pending",
  "running",
  "completed",
  "failed",
  "cancelled",
];

export const WORKFLOW_STEP_STATUSES: WorkflowStepStatus[] = [
  "pending",
  "running",
  "completed",
  "failed",
  "skipped",
];

export interface WorkflowDefinition {
  id: string;
  name: string;
  description: string | null;
  triggerType: WorkflowTriggerType;
  enabled: boolean;
  organizationId: string | null;
  version: number;
  metadata: Record<string, unknown>;
  createdAt: string;
  updatedAt: string;
}

export interface WorkflowDefinitionInput {
  name: string;
  description?: string | null;
  triggerType: WorkflowTriggerType;
  enabled?: boolean;
  organizationId?: string | null;
  version?: number;
  metadata?: Record<string, unknown>;
}

export interface WorkflowTrigger {
  id: string;
  workflowDefinitionId: string;
  eventName: string;
  eventSource: string | null;
  conditionsJson: Record<string, unknown>;
  metadata: Record<string, unknown>;
  createdAt: string;
  updatedAt: string;
}

export interface WorkflowTriggerInput {
  eventName: string;
  eventSource?: string | null;
  conditionsJson?: Record<string, unknown>;
  metadata?: Record<string, unknown>;
}

export interface WorkflowAction {
  id: string;
  workflowDefinitionId: string;
  sequence: number;
  actionType: string;
  configurationJson: Record<string, unknown>;
  enabled: boolean;
  metadata: Record<string, unknown>;
  createdAt: string;
  updatedAt: string;
}

export interface WorkflowActionInput {
  sequence?: number;
  actionType: string;
  configurationJson?: Record<string, unknown>;
  enabled?: boolean;
  metadata?: Record<string, unknown>;
}

export interface WorkflowExecution {
  id: string;
  workflowDefinitionId: string;
  triggerEvent: string | null;
  status: WorkflowStatus;
  startedAt: string | null;
  completedAt: string | null;
  failureReason: string | null;
  contextJson: Record<string, unknown>;
  metadata: Record<string, unknown>;
  createdAt: string;
  updatedAt: string;
}

export interface WorkflowExecutionStep {
  id: string;
  workflowExecutionId: string;
  workflowActionId: string;
  status: WorkflowStepStatus;
  startedAt: string | null;
  completedAt: string | null;
  retryCount: number;
  outputJson: Record<string, unknown>;
  errorMessage: string | null;
  metadata: Record<string, unknown>;
  createdAt: string;
  updatedAt: string;
}

export interface WorkflowDefinitionListQuery {
  q?: string;
  triggerType?: WorkflowTriggerType;
  enabled?: boolean;
  organizationId?: string;
  createdFrom?: string;
  createdTo?: string;
  page?: number;
  pageSize?: number;
}

export interface WorkflowExecutionListQuery {
  status?: WorkflowStatus;
  startedFrom?: string;
  startedTo?: string;
  page?: number;
  pageSize?: number;
}

export interface WorkflowStepListQuery {
  status?: WorkflowStepStatus;
  page?: number;
  pageSize?: number;
}

export interface WorkflowDashboardStats {
  totalWorkflows: number;
  enabledWorkflows: number;
  runningExecutions: number;
  failedExecutions: number;
  completedExecutions: number;
  recentExecutions: WorkflowExecution[];
}

// ─── Phase 8.5 — Runtime monitoring ─────────────────────────────────────

export type RuntimeHealthStatus = "ok" | "degraded" | "unhealthy" | "unknown";

export interface RuntimeHealthCheck {
  status: RuntimeHealthStatus;
  detail?: string;
  [key: string]: unknown;
}

export interface RuntimeHealth {
  status: RuntimeHealthStatus;
  checks: {
    registry: RuntimeHealthCheck;
    scheduler: RuntimeHealthCheck;
    queue: RuntimeHealthCheck;
    celery: RuntimeHealthCheck;
    handlers: RuntimeHealthCheck;
  };
}

export interface RuntimeStatisticsOverview {
  total: number;
  byStatus: Record<string, number>;
  completed: number;
  failed: number;
  cancelled: number;
  running: number;
  successRate: number;
  failureRate: number;
  retryRate: number;
  avgDurationSeconds: number;
  totalRetries: number;
  retryExecutions: number;
  since: string | null;
}

export interface RuntimeTopWorkflow {
  workflowDefinitionId: string;
  name: string | null;
  total: number;
}

export interface RuntimeTopFailure {
  workflowDefinitionId: string;
  name: string | null;
  failed: number;
}

export interface RuntimeMetricsAggregate {
  count: number;
  total: number;
  average: number;
  min: number | null;
  max: number | null;
}

export interface RuntimeMetricsSnapshot {
  executionsTotal: number;
  executionsByStatus: Record<string, number>;
  executionsByWorkflow: Record<string, number>;
  duration: RuntimeMetricsAggregate;
  handlerDuration: Record<string, RuntimeMetricsAggregate>;
  queueLatency: RuntimeMetricsAggregate;
  retryCount: number;
  actionSuccess: number;
  actionFailure: number;
  generatedAt: string;
}

export interface RuntimeStatistics {
  overview: RuntimeStatisticsOverview;
  topWorkflows: RuntimeTopWorkflow[];
  topFailures: RuntimeTopFailure[];
  metrics: RuntimeMetricsSnapshot;
}

export interface RetryHistoryStep {
  stepId: string;
  actionId: string;
  attempt: number;
  retryCount: number;
  status: WorkflowStepStatus;
  finalStatus: WorkflowStepStatus | null;
  lastError: string | null;
  startedAt: string | null;
  completedAt: string | null;
}

export interface RetryHistory {
  executionId: string;
  workflowDefinitionId: string;
  status: WorkflowStatus;
  totalSteps: number;
  totalRetries: number;
  steps: RetryHistoryStep[];
}

File: backend/app/schemas/workflow.py
"""Pydantic schemas for the Automation & Workflow Engine (Phase 7.1)."""
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.schemas.common import IdentifiedDto


TriggerType = Literal["event", "schedule", "manual"]
WorkflowStatus = Literal["pending", "running", "completed", "failed", "cancelled"]
StepStatus = Literal["pending", "running", "completed", "failed", "skipped"]
ActionType = Literal[
    "notification",
    "audit",
    "analytics",
    "webhook",
    "email",
    "sms",
    "update_entity",
    "custom",
]


# --------------------------------------------------------------------------- #
# WorkflowDefinition
# --------------------------------------------------------------------------- #


class WorkflowDefinitionBase(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    name: str = Field(min_length=1, max_length=200)
    description: str | None = None
    triggerType: TriggerType = Field(alias="trigger_type")
    enabled: bool = True
    organizationId: uuid.UUID | None = Field(alias="organization_id", default=None)
    version: int = Field(default=1, ge=1)


class WorkflowDefinitionCreate(WorkflowDefinitionBase):
    metadata: dict[str, Any] = Field(default_factory=dict)


class WorkflowDefinitionUpdate(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    name: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = None
    triggerType: TriggerType | None = Field(alias="trigger_type", default=None)
    enabled: bool | None = None
    version: int | None = Field(default=None, ge=1)
    metadata: dict[str, Any] | None = None


class WorkflowDefinitionDto(IdentifiedDto):
    name: str
    description: str | None = None
    triggerType: TriggerType = Field(alias="trigger_type")
    enabled: bool
    organizationId: uuid.UUID | None = Field(alias="organization_id", default=None)
    version: int
    metadata: dict[str, Any] = Field(default_factory=dict)


class WorkflowDefinitionListQuery(BaseModel):
    triggerType: TriggerType | None = None
    enabled: bool | None = None
    organizationId: uuid.UUID | None = None
    page: int = 1
    pageSize: int = 50


class WorkflowSearchQuery(BaseModel):
    q: str | None = None
    triggerType: TriggerType | None = None
    status: WorkflowStatus | None = None
    page: int = 1
    pageSize: int = 50


# --------------------------------------------------------------------------- #
# WorkflowTrigger
# --------------------------------------------------------------------------- #


class WorkflowTriggerBase(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    workflowDefinitionId: uuid.UUID = Field(alias="workflow_definition_id")
    eventName: str = Field(alias="event_name", min_length=1, max_length=120)
    eventSource: str | None = Field(alias="event_source", default=None, max_length=120)
    conditionsJson: dict[str, Any] = Field(
        alias="conditions_json", default_factory=dict
    )


class WorkflowTriggerCreate(WorkflowTriggerBase):
    metadata: dict[str, Any] = Field(default_factory=dict)


class WorkflowTriggerUpdate(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    eventName: str | None = Field(alias="event_name", default=None, max_length=120)
    eventSource: str | None = Field(alias="event_source", default=None, max_length=120)
    conditionsJson: dict[str, Any] | None = Field(alias="conditions_json", default=None)
    metadata: dict[str, Any] | None = None


class WorkflowTriggerDto(IdentifiedDto):
    workflowDefinitionId: uuid.UUID = Field(alias="workflow_definition_id")
    eventName: str = Field(alias="event_name")
    eventSource: str | None = Field(alias="event_source", default=None)
    conditionsJson: dict[str, Any] = Field(alias="conditions_json", default_factory=dict)
    metadata: dict[str, Any] = Field(default_factory=dict)


# --------------------------------------------------------------------------- #
# WorkflowAction
# --------------------------------------------------------------------------- #


class WorkflowActionBase(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    workflowDefinitionId: uuid.UUID = Field(alias="workflow_definition_id")
    sequence: int = Field(ge=1)
    actionType: ActionType = Field(alias="action_type")
    configurationJson: dict[str, Any] = Field(
        alias="configuration_json", default_factory=dict
    )
    enabled: bool = True


class WorkflowActionCreate(WorkflowActionBase):
    metadata: dict[str, Any] = Field(default_factory=dict)


class WorkflowActionUpdate(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    sequence: int | None = Field(default=None, ge=1)
    actionType: ActionType | None = Field(alias="action_type", default=None)
    configurationJson: dict[str, Any] | None = Field(
        alias="configuration_json", default=None
    )
    enabled: bool | None = None
    metadata: dict[str, Any] | None = None


class WorkflowActionDto(IdentifiedDto):
    workflowDefinitionId: uuid.UUID = Field(alias="workflow_definition_id")
    sequence: int
    actionType: ActionType = Field(alias="action_type")
    configurationJson: dict[str, Any] = Field(
        alias="configuration_json", default_factory=dict
    )
    enabled: bool
    metadata: dict[str, Any] = Field(default_factory=dict)


# --------------------------------------------------------------------------- #
# WorkflowExecution
# --------------------------------------------------------------------------- #


class WorkflowExecutionBase(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    workflowDefinitionId: uuid.UUID = Field(alias="workflow_definition_id")
    triggerEvent: str | None = Field(alias="trigger_event", default=None, max_length=120)
    status: WorkflowStatus = "pending"
    contextJson: dict[str, Any] = Field(alias="context_json", default_factory=dict)

    @model_validator(mode="after")
    def _validate_period(self):
        return self


class WorkflowExecutionCreate(WorkflowExecutionBase):
    startedAt: datetime | None = Field(alias="started_at", default=None)
    completedAt: datetime | None = Field(alias="completed_at", default=None)
    failureReason: str | None = Field(alias="failure_reason", default=None)
    metadata: dict[str, Any] = Field(default_factory=dict)

    @model_validator(mode="after")
    def _validate_completion(self):
        if (
            self.startedAt is not None
            and self.completedAt is not None
            and self.completedAt < self.startedAt
        ):
            raise ValueError("completed_at must be >= started_at")
        return self


class WorkflowExecutionUpdate(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    status: WorkflowStatus | None = None
    startedAt: datetime | None = Field(alias="started_at", default=None)
    completedAt: datetime | None = Field(alias="completed_at", default=None)
    failureReason: str | None = Field(alias="failure_reason", default=None)
    contextJson: dict[str, Any] | None = Field(alias="context_json", default=None)
    metadata: dict[str, Any] | None = None


class WorkflowExecutionDto(IdentifiedDto):
    workflowDefinitionId: uuid.UUID = Field(alias="workflow_definition_id")
    triggerEvent: str | None = Field(alias="trigger_event", default=None)
    status: WorkflowStatus
    startedAt: datetime | None = Field(alias="started_at", default=None)
    completedAt: datetime | None = Field(alias="completed_at", default=None)
    failureReason: str | None = Field(alias="failure_reason", default=None)
    contextJson: dict[str, Any] = Field(alias="context_json", default_factory=dict)
    metadata: dict[str, Any] = Field(default_factory=dict)


class WorkflowExecutionListQuery(BaseModel):
    workflowDefinitionId: uuid.UUID | None = None
    status: WorkflowStatus | None = None
    startedFrom: datetime | None = None
    startedTo: datetime | None = None
    page: int = 1
    pageSize: int = 50


# --------------------------------------------------------------------------- #
# WorkflowExecutionStep
# --------------------------------------------------------------------------- #


class WorkflowExecutionStepBase(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    workflowExecutionId: uuid.UUID = Field(alias="workflow_execution_id")
    workflowActionId: uuid.UUID = Field(alias="workflow_action_id")
    status: StepStatus = "pending"
    retryCount: int = Field(alias="retry_count", default=0, ge=0)


class WorkflowExecutionStepCreate(WorkflowExecutionStepBase):
    startedAt: datetime | None = Field(alias="started_at", default=None)
    completedAt: datetime | None = Field(alias="completed_at", default=None)
    outputJson: dict[str, Any] = Field(alias="output_json", default_factory=dict)
    errorMessage: str | None = Field(alias="error_message", default=None)
    metadata: dict[str, Any] = Field(default_factory=dict)


class WorkflowExecutionStepUpdate(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    status: StepStatus | None = None
    startedAt: datetime | None = Field(alias="started_at", default=None)
    completedAt: datetime | None = Field(alias="completed_at", default=None)
    retryCount: int | None = Field(alias="retry_count", default=None, ge=0)
    outputJson: dict[str, Any] | None = Field(alias="output_json", default=None)
    errorMessage: str | None = Field(alias="error_message", default=None)
    metadata: dict[str, Any] | None = None


class WorkflowExecutionStepDto(IdentifiedDto):
    workflowExecutionId: uuid.UUID = Field(alias="workflow_execution_id")
    workflowActionId: uuid.UUID = Field(alias="workflow_action_id")
    status: StepStatus
    startedAt: datetime | None = Field(alias="started_at", default=None)
    completedAt: datetime | None = Field(alias="completed_at", default=None)
    retryCount: int = Field(alias="retry_count")
    outputJson: dict[str, Any] = Field(alias="output_json", default_factory=dict)
    errorMessage: str | None = Field(alias="error_message", default=None)
    metadata: dict[str, Any] = Field(default_factory=dict)
